### DISCUSSION TITLE
IP_FORWARDING

### DISCUSSION THREAD

**Discussion by skirmess** [FoobarOy/foomuuri#40] [Ideas]
[Shorewall](https://shorewall.org/manpages/shorewall.conf.html) has the IP_FORWARDING option. This option ensures that the firewall is not forwarding packets before the rule is successfully loaded. I do think that with IPv6 it becomes even more important to not forward any packets before the rules are loaded. Otherwise the system would be vulnerable whenever the firewall reboots or if the rules aren't successfully loaded - for whatever reason - after a reboot.

Would that be something that Foomuuri could/should be able to control? Or what is the best practice to securely reboot a firewall with Foomuuri?

> **IP_FORWARDING=[On|Off|Keep]**
>
>     This IPv4 parameter determines whether Shorewall enables or disables IPv4 Packet Forwarding (/proc/sys/net/ipv4/ip_forward). In an IPv6 configuration, this parameter determines the setting of /proc/sys/net/ipv6/config/all/ip_forwarding.
>
>     Possible values are:
>
>     On or on
>         packet forwarding will be enabled.
>     Off or off
>         packet forwarding will be disabled.
>     Keep or keep
>         Shorewall will neither enable nor disable packet forwarding.
>
>     If this variable is not set or is given an empty value (IP_FORWARD="") then IP_FORWARD=On is assumed.

**Comment by rfrancoise** [comment]
On systemd-based systems, the foomuuri service has:

```
Before=network-pre.target
Wants=network-pre.target
```

ensuring that the firewall starts before the network even comes up, meaning no traffic can pass before the rules are loaded.

DNS and iplist sets configured in the `resolve{}` and `iplist{}` sections are cached on disk to make them available during early boot.

**Comment by skirmess** [comment]
What happens (sorry, I've never really used systemd) if the firewall doesn't start (corrupt rule file, foomuuri binary can't execute (disk problem), bug in the code)? Does systemd not bring up the network at all?

**Comment by kimheino** [comment]
If config file is corrupted Foomuuri will fallback to previous "good" config, loading at least some firewall.

If Foomuuri binary is corrupted network will come up without firewall.

If you are afraid of packets being forwarded you can use Foomuuri's `hook { post_start /path/to/script/sysctl.enable.forwarding }`. It will be executed after Foomuuri has loaded the rules. See https://github.com/FoobarOy/foomuuri/wiki/Configuration#hook

**Reply by rfrancoise** [reply]
Similarly to Shorewall, Foomuuri could ship a `foomuuri-init` systemd unit which would install a minimal default-deny policy and start before the normal service. Upon normal startup this policy would then get replaced and if anything were to go wrong, the system would remain protected (but unreachable).

**Reply by bleve** [reply]
Foomuuri has good.fw saved especially for this reason. If configuration has been changed in a way it can't be loaded just before reboot and loading current configuration fails, foomuuri loads good.fw which is created from latest successfully loaded configuration.

**Reply by kimheino** [reply]
Minimal "default-drop policy" binary to be started very early on boot, before Foomuuri:

```
#!/usr/sbin/nft -f

table inet foomuuri
delete table inet foomuuri

table inet foomuuri {
        chain input {
                type filter hook input priority filter
		drop
	}

        chain output {
                type filter hook output priority filter
		drop
	}

        chain forward {
                type filter hook forward priority filter
		drop
	}
}
```

And as bleve said, "good.fw" aka " fallback to previous "good" config" should make this unneeded. Therefore I'm not including it, at least not yet.

**Reply by rfrancoise** [reply]
I'm aware that Foomuuri saves the last good ruleset that Nftables accepted. However, I don't think it'll help if Python modules are hosed in a way that prevents Foomuuri from running, or even if a syntax error has been unknowingly added to `foomuuri.conf` since the last boot. In that case Foomuuri will exit and not load anything.

**Reply by kimheino** [reply]
Ah, you are right. Syntax error in config can result bootup without firewall. In cases like this, what do you prefer:

1) Drop-all firewall, no access to system.
2) Last good ruleset.

Option 2 can be archived with simple service file:

Before=foomuuri.service
ExecStart=nft -f /var/lib/foomuuri/good.fw

It could be implement inside Foomuuri too, but broken python modules would still break it.

**Reply by rfrancoise** [reply]
I was thinking about this after posting my message last night and an even simpler solution would be to add this to the standard service:
```
ExecStartPre=-/sbin/nft -f /var/lib/foomuuri/good.fw
```
The `-` will ensure that any Nftables issues don't block startup and result in a proper Foomuuri error.

But maybe some people would still prefer a deny-all defaut policy like we had with `shorewall-init`.

@skirmess WDYT?

**Reply by kimheino** [reply]
Personally I prefer "good.fw" much more than a locked out system. It might be really hard to fix such system. "good.fw" is usually good enough.

Do you think that "broken python" is really a case? I'm thinking a solution that good.fw-loader would be moved to "syntax error / nft error" -handler. This would ensure:

- Normally Foomuuri loads current config, and only that
- If syntax error, good.fw will be loaded (only if command was "start-or-good")
- No double-loading
- No locked out system
- No extra delays (yeah, few milliseconds)

Config option / command line "start-or-block" to select between good.fw and block-all can be added later.

**Reply by kimheino** [reply]
`ExecStartPre` is not ideal solution, as that would result "current ruleset -> good.fw -> new ruleset" transition in `systemctl restart foomuuri`. I prefer atomic switches.

That's why separate `foomuuri-reboot.service` would be better. Jus tested, large but not huge good.fw is loaded in ~300ms.

**Reply by rfrancoise** [reply]
Foomuuri imports some modules with complex dependency chains (dbus, glib). Those could fail to be imported if the system reboots in the middle of a system upgrade, for example. A rare problem, but a plausible one...

It won't hurt to make the exception handler attempt to restore `good.fw` when possible, in any case. We can do that and also provide a separate "boot/init" systemd unit for more paranoid users.

And good point about `ExecStartPre`, indeed that wouldn't be a good outcome. Thanks for pointing it out.

**Reply by skirmess** [reply]
This being a firewall it's better to be paranoid, twice. A security system should always fail save.

Worst case scenario I can think of is a power failure, during my vacation, that reboots the system but doesn't load the rule base, leaving the network open and, thanks to IPv6*, every system fully accessible.

*) Yes, NAT is not a security feature but having a public address exposes the systems without any effort if the firewall doesn't load a rule base.

Things I would like to see handled:

- Corruption in any config file

Could be that I've edited a config file but didn't restart foomuuri and forgot about this. If the config file is now corrupt the rules might not reload after rebooting.

Maybe the change was in a config file not used during restart, only during start and the issue therefore only occurs during reboot.

Maybe I've updated the packages on my system, including foomuuri, and forgot to restart foomuuri. A change in foomuuri could make the old config, or rule file invalid, either though a bug, a former undefned behaviour, or through a breaking change.

Solution: The system should, after rebooting, not forward packages if foomuuri did not successfully run.

Ideally a foomuuri-init would run during boot and load a deny policy that blocks everything besides what's needed (DHCP client) and SSH from a defined zone.

Additionally, IP forwarding should be disabled in the system and only enable by foomuuri after the firewall is enabled correctly. (I guess that could be done with the hook)

- Broken python libraries or python interpreter

A Python library could stop working because of a change of a library, or a dependency of a library.

The Python interpreter could stop working because of a system upgrade that changed/broke a dependency of Python. Or foomuuri could stop working because of a Python upgrade that broke compatibility.

Solution: The foomuuri-init, or the good.fw rulebase, should be loaded without requiring Python. (I guess this would be the case if we load good.fw with nft from systemd?)

- nftables rule doesn't load because of change in kernel

Not the most likely occurence but maybe a complicated good.fw rule isn't good anymore after a change to the kernel. A small default deny rule loaded first is much less likely to break because of these changes.

Solution: Initialize system with a simple, easy rule during early boot and don't enable forwarding before the real rule is loaded.

- good.fw can't be read because of read error (bad block on disk, file system corruption)

Disks do break and often firewalls don't use raid. In the worst possible situation the system would still be able to boot but the good.fw file couldn't be loaded.

There could also be a bug in the file system driver that doesn't allow the file to be read. Or maybe the file was changed to `chmod 000 good.fw`.


Solution: Don't enable forwarding unless the rule is successfully loaded

**Comment by kimheino** [comment]
Here's my current plan:

Problem 1: Boot without firewall if Python is broken / can't run Foomuuri.

Solution 1: Add `foomuuri-boot.service` that is run before `foomuuri.service` on bootup. It loads `good.fw` (or `drop-all.fw` or whatever, your choice) with nft binary. This is run before network-pre.target so firewall is up before your network is up.

Problen 2: Syntax error in config.

Solution 2: Already solved by solution 1. Side effect: `foomuuri start-or-good` command line can be removed.

Problem 3: IP forwarding.

Solution 3: Use Foomuuri's `post_start` hook to enable IP forwarding, not `/etc/sysctl.d/`. Foomuuri doesn't touch these settings, similar to Shorewall's `IP_FORWARDING=Keep`.


And some comments:

> Solution: The system should, after rebooting, not forward packages if foomuuri did not successfully run.

Solution 3 above.

> Ideally a foomuuri-init would run during boot and load a deny policy that blocks everything besides what's needed (DHCP client) and SSH from a defined zone.

Solution 1 above.

> Solution: The foomuuri-init, or the good.fw rulebase, should be loaded without requiring Python. (I guess this would be the case if we load good.fw with nft from systemd?)

Solution 1 above, loaded with `nft` binary.

> nftables rule doesn't load because of change in kernel

Solution 1, with drop-all or similar minimal ruleset. Not likely to happen.

> Disks do break and often firewalls don't use raid.

Firewall must use RAID. Disk failure is much more likely than kernel change.

> Solution: Don't enable forwarding unless the rule is successfully loaded

Solution 3.

I'll add only two stages: boot.service to load good/deny/minimal, foomuuri.service to load current. Three stages (deny -> good -> current) is overkill as all this happens before your network is up.

**Reply by kimheino** [reply]
Commits a4c7f75 and 06d4109 implements solution 1.

**Reply by kimheino** [reply]
Commit cc83e43 removes `start-or-good`

**Comment by rfrancoise** [comment]
Works very nicely in v0.24, thank you. I'll sleep easier knowing that if my `foomuuri.conf` is corrupted somehow, my last good firewall policy will still come up on the next reboot. 👍
